### PR TITLE
Hidden scripts

### PR DESCRIPTION
**Pull request by Bjvanminnen**
It is currently the case on our site that teachers are able (in certain scripts) to hide stages on a per section basis. Going forward, we also want them to be able to hide scripts within a course on a per section basis.

This PR does a lot of the prep work for this. In its current state, it should result in no behavior difference, as there is no way one to create a SectionHiddenScript yet. However, if you were to create one manually via the db, the behavior for students when going to a puzzle page in that script would be the same as when you go to a puzzle page in a hidden stage.

Future work will involve making it so that the script is hidden on the course overview page, and that teachers actually have UI to toggle the hide state.

### CODE REVIEW

File: dashboard/app/controllers/script_levels_controller.rb [hunk 1]
```diff
@@ -369,40 +374,6 @@ def present_level
...
-      all_ids = script_sections.flat_map(&:section_hidden_stages).pluck(:stage_id)
-      counts = all_ids.each_with_object(Hash.new(0)) {|id, hash| hash[id] += 1}
-      counts.select {|_, val| val == script_sections.length}.keys
-    else
-      # if we have no sections matching this script id, we consider a stage hidden if any of those sections
-      # hides the stage
-      sections.flat_map(&:section_hidden_stages).pluck(:stage_id).uniq
-    end
-  end
-
```
Comment (Bjvanminnen) [inline dashboard/app/controllers/script_levels_controller.rb:405]: This was all moved into the User model

File: dashboard/app/models/user.rb [hunk 2]
```diff
@@ -876,14 +877,81 @@ def hidden_stage?(script_level)
...
+      if hidden_stages
+        all_ids = assigned_sections.flat_map(&:section_hidden_stages).pluck(:stage_id)
+      else
+        all_ids = assigned_sections.flat_map(&:section_hidden_scripts).pluck(:script_id)
+      end
+
+      counts = all_ids.each_with_object(Hash.new(0)) {|id, hash| hash[id] += 1}
+      counts.select {|_, val| val == assigned_sections.length}.keys
+    end
+  end
```
Comment (Bjvanminnen) [inline dashboard/app/models/user.rb]: I was somewhat torn on how to best write this method.

I originally considered taking the initial `hidden_stage?` method, duplicating it, and then changing it to find scripts instead of stages as appropriate. This however was a lot of duplicated code.

I then explored passing in the map function (i.e. section_hidden_stages vs section_hidden_scripts), and other data (i.e. :stage_id vs. script_id) as params. I never got this fully working, but could have if I'd stuck with it. However, the resulting method felt very difficult to read.

I ultimately settled on this approach of having a bunch of places where we gate on `hidden_stages` or not. I don't love it, but I prefer it to the alternatives.

File: dashboard/test/controllers/script_levels_controller_test.rb [hunk 3]
```diff
@@ -1288,162 +1288,18 @@ def assert_caching_enabled(cache_control_header, max_age, proxy_max_age)
...
-    SectionHiddenStage.create(section_id: teacher_member_section.id, stage_id: stage3.id)
-
-    response = get_hidden(@script)
-    hidden = JSON.parse(response.body)
-    # only the stages hidden in the owned section are considered hidden
-    expected = {
-      teacher_owner_section.id.to_s => [stage1.id],
-      teacher_owner_section2.id.to_s => [stage1.id, stage2.id]
-    }
-    assert_equal expected, hidden
```
Comment (Bjvanminnen) [inline dashboard/test/controllers/script_levels_controller_test.rb:1446]: moved a whole chunk of these tests over to user_tests

File: dashboard/app/models/section_hidden_script.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+# Indexes
+#
+#  index_section_hidden_scripts_on_script_id   (script_id)
+#  index_section_hidden_scripts_on_section_id  (section_id)
+#
+
+class SectionHiddenScript < ApplicationRecord
+  belongs_to :section
+  belongs_to :script
+end
```
Comment (ashercodeorg) [inline dashboard/app/models/section_hidden_script.rb:18]: Please keep the migration to its own PR, deployed separate from any references or usages of the model.

Comment (Bjvanminnen) [reply dashboard/app/models/section_hidden_script.rb:18]: It is...

Comment (Bjvanminnen) [reply dashboard/app/models/section_hidden_script.rb:18]: Migration was https://github.com/code-dot-org/code-dot-org/pull/17261

Comment (ashercodeorg) [reply dashboard/app/models/section_hidden_script.rb:18]: Ah, so this PR just needs rebasing. Never mind, ignore me...

File: dashboard/test/models/user_test.rb [hunk 5]
```diff
@@ -2186,4 +2186,225 @@ class SectionCourses < ActiveSupport::TestCase
...
+      SectionHiddenScript.create(section_id: teacher_member_section.id, script_id: @script3.id)
+
+      # only the scripts hidden in the owned section are considered hidden
+      expected = {
+        teacher_owner_section.id => [@script.id],
+        teacher_owner_section2.id => [@script.id, @script2.id]
+      }
+      assert_equal expected, teacher.get_hidden_script_ids(@course)
+    end
+  end
```
Comment (Bjvanminnen) [inline dashboard/test/models/user_test.rb:2409]: This are mostly tests that had existed elsewhere, but cleaned up a little bit to repeat less. I then duplicated them to have the same logic we used to check for hidden stages in a script to instead check for hidden scripts in a course.

File: dashboard/app/controllers/script_levels_controller.rb (hunk 1 above)
Comment (islemaster) [reply dashboard/app/controllers/script_levels_controller.rb:405]: :clap: +1 to fat model, thin controller!

File: dashboard/app/models/ability.rb [hunk 6]
```diff
@@ -57,6 +58,7 @@ def initialize(user)
       can :read, UserPermission, user_id: user.id
       can [:show, :pull_review, :update], PeerReview, reviewer_id: user.id
       can :read, SectionHiddenStage
+      can :read, SectionHiddenScript
```
Comment (islemaster) [inline dashboard/app/models/ability.rb]: I don't follow where this permission gets checked.  Is it implicit in a `load_and_authorize_resource` call somewhere?

File: dashboard/app/models/script_level.rb [hunk 7]
```diff
@@ -176,7 +176,7 @@ def valid_progression_level?(user=nil)
 
   def locked_or_hidden?(user)
     return false unless user
-    return true if user.hidden_stage?(self)
+    return true if user.script_level_hidden?(self)
     return true if locked?(user)
     false
   end
```
Comment (islemaster) [inline dashboard/app/models/script_level.rb]: Why not collapse this method to:
```ruby
def locked_or_hidden?(user)
  user && (user.script_level_hidden?(self) || locked?(user))
end
```

File: dashboard/app/models/user.rb [hunk 8]
```diff
@@ -876,14 +877,81 @@ def hidden_stage?(script_level)
...
+      script_sections.all? {|s| script_level.hidden_for_section?(s.id)}
     else
       # if we have no sections matching this script id, we consider a stage hidden if any of the sections we're in
       # hide it
-      sections.any? {|s| script_level.stage_hidden_for_section?(s.id)}
+      sections.any? {|s| script_level.hidden_for_section?(s.id)}
     end
   end
 
+  # This method will extra a list of hidden ids by section. The type of ids depends
```
Comment (islemaster) [inline dashboard/app/models/user.rb]: `s/extra/extract`?

File: dashboard/app/models/user.rb [hunk 9]
```diff
@@ -876,14 +877,81 @@ def hidden_stage?(script_level)
...
+      hidden_stages && section.script_id == assign_id || !hidden_stages && section.course_id == assign_id
+    end
+
+    if assigned_sections.empty?
+      # if we have no sections matching this assignment, we consider a stage/script
+      # hidden if any of our sections hides it
+      if hidden_stages
+        sections.flat_map(&:section_hidden_stages).pluck(:stage_id).uniq
+      else
+        sections.flat_map(&:section_hidden_scripts).pluck(:script_id).uniq
```
Comment (islemaster) [inline dashboard/app/models/user.rb]: It took me a minute to realize that lines 923 and 925 here are return values, being the last statement to execute in this path through this method even though there's 10+ lines left.  If our linter will allow it, it'd be nice to list explicit returns or assign to a local and return that at the end, to avoid confusion.

And/or, these five lines are largely duplicated with the ones in the else case below, only different by which section list we pass in and what we do with the result.  Might be a good place for a private helper.

File: dashboard/app/models/user.rb [hunk 10]
```diff
@@ -876,14 +877,81 @@ def hidden_stage?(script_level)
...
+      sections.each do |section|
+        if hidden_stages
+          # Given a script_id, looking for hidden stages
+          hidden_by_section[section.id] = section.section_hidden_stages.pluck(:stage_id)
+        else
+          # Given a course_id, looking for hidden_scripts
+          hidden_by_section[section.id] = section.section_hidden_scripts.pluck(:script_id)
+        end
+      end
+      return hidden_by_section
```
Comment (islemaster) [inline dashboard/app/models/user.rb]: You've got two totally non-overlapping cases in teacher/student here.  Maybe break out private helpers `get_hidden_ids_for_teacher` `get_hidden_ids_for_student`?

File: dashboard/test/controllers/script_levels_controller_test.rb [hunk 11]
```diff
@@ -1288,162 +1288,18 @@ def assert_caching_enabled(cache_control_header, max_age, proxy_max_age)
...
-  def put_student_in_section(student, teacher, script)
-    section = create :section, user_id: teacher.id, script_id: script.id
-    Follower.create!(section_id: section.id, student_user_id: student.id, user: teacher)
-    section
-  end
-
-  def get_hidden(script)
-    response = get :hidden, params: {script_id: script.name}
+  test "hidden_stage_ids for user not signed in" do
+    response = get :hidden_stage_ids, params: {script_id: @script.name}
```
Comment (islemaster) [inline dashboard/test/controllers/script_levels_controller_test.rb:1292]: I'd add a test for expected behavior if someone tries to call the `hidden_stage_ids` route without the `script_id` param.

File: dashboard/test/factories/factories.rb [hunk 12]
```diff
@@ -7,7 +7,7 @@
   end
 
   factory :course do
-    name "my-course-name"
+    sequence(:name) {|n| "bogus-course-#{n}"}
```
Comment (islemaster) [inline dashboard/test/factories/factories.rb:10]: qvy8d_s-200x150

File: dashboard/test/models/script_test.rb [hunk 13]
```diff
@@ -267,8 +267,10 @@ def populate_cache_and_disconnect_db
       refute s.hidden?, "#{s.name} is hidden when it should not be"
     end
 
+    # filter out factory created scripts
+    all_scripts = Script.all.select {|s| !(/^bogus-script.*/ =~ s.name)}
```
Comment (islemaster) [inline dashboard/test/models/script_test.rb]: This feels like possible spooky action at a distance.  Can we share a constant with `factories.rb`, or somehow define these side-by-side?

Comment (Bjvanminnen) [reply dashboard/test/models/script_test.rb]: I'd asked about this test in dashboard-dev, but didn't get any response.

I *think* the entire test is actually useless. It was originally created for us to catch when content writers accidentally toggled the hidden state of a script. These days, we get these all from fixtures (and apparently factory-created scripts, depending on when you create them). As such, I don't think the test gives us any value, and I should probably just remove the entire thing.

Agree/disagree?

Comment (islemaster) [reply dashboard/test/models/script_test.rb]: We're... using tests to catch changes to our fixtures?  Sure, cut it.

If we want to test which scripts are hidden, we should probably do that as late as possible in an acceptance (UI) test.

File: dashboard/app/models/ability.rb (hunk 6 above)
Comment (Bjvanminnen) [reply dashboard/app/models/ability.rb]: You might be right that this is unnecessary. There is an existing test for SectionHiddenStage, but I think it's just validating this ability - not its usage.

When toggling hidden stages, we make sure we have the ability to read the section (https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/controllers/script_levels_controller.rb#L122) but we don't check SectionHiddenStage. I suspect I can remove all ability code for SectionHiddenStage/Script, as well as the relevant test.

File: dashboard/app/models/user.rb (hunk 8 above)
Comment (aoby) [inline dashboard/app/models/user.rb]: extra?

File: dashboard/app/models/user.rb (hunk 10 above)
Comment (aoby) [reply dashboard/app/models/user.rb]: Agreed, this would be more readable.

File: dashboard/app/models/user.rb [hunk 14]
```diff
@@ -876,14 +877,81 @@ def hidden_stage?(script_level)
...
+    if teacher?
+      hidden_by_section = {}
+      sections.each do |section|
+        if hidden_stages
+          # Given a script_id, looking for hidden stages
+          hidden_by_section[section.id] = section.section_hidden_stages.pluck(:stage_id)
+        else
+          # Given a course_id, looking for hidden_scripts
+          hidden_by_section[section.id] = section.section_hidden_scripts.pluck(:script_id)
+        end
```
Comment (aoby) [inline dashboard/app/models/user.rb]: Can this be combined with the similar mapping below? Either in a helper called in both places, or collect the relevant sections for teacher / student, then map stage or script ids. They look very similar, but this constructs a hash and the other uses a unique flat map. Does this mean the function returns either an array or hash depending on user type? Can they be consolidated? At least add a `@returns` function header explaining that.

File: dashboard/test/controllers/script_levels_controller_test.rb (hunk 11 above)
Comment (Bjvanminnen) [reply dashboard/test/controllers/script_levels_controller_test.rb:1292]: The param is part of the path, so I'm not sure that's possible, i.e. the route is something like `/s/csp1/hidden_stages`

File: dashboard/app/models/user.rb (hunk 14 above)
Comment (Bjvanminnen) [reply dashboard/app/models/user.rb]: This function does return either an array or hash, which I've never really liked. Making this perhaps even sillier, the client - upon getting back an array - essentially turns it into a hash with a pretend section id of "STUDENT". https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/code-studio/hiddenStageRedux.js#L109

I could just as easily have the server always return a hash in this form. It would mean our fake STUDENT section id would now need to be shared between server/client, which feels gross, but maybe no grosser than what we have already.

File: dashboard/test/controllers/script_levels_controller_test.rb (hunk 11 above)
Comment (islemaster) [reply dashboard/test/controllers/script_levels_controller_test.rb:1292]: My bad, I misunderstood the route. :+1:

File: dashboard/app/models/user.rb (hunk 14 above)
Comment (Bjvanminnen) [reply dashboard/app/models/user.rb]: After starting to write this code, it still actually feels cleaner to me to have the server API return a different shape of data for teachers vs. students (and document this) then to have a magic section id that is not really a section id just to avoid this.

Comment (aoby) [reply dashboard/app/models/user.rb]: I'm ok with that. Currently (without documentation), it's unexpected and confusing.

File: dashboard/app/models/user.rb [hunk 15]
```diff
@@ -876,14 +877,92 @@ def hidden_stage?(script_level)
...
+    else
+      # if we do have sections matching this assignment, we consider a stage/script
+      # hidden only if it is hidden in every one of the sections the student belongs
+      # to that match this assignment
+      all_ids = hidden_stages ? hidden_stage_ids(assigned_sections) : hidden_script_ids(assigned_sections)
+
+      counts = all_ids.each_with_object(Hash.new(0)) {|id, hash| hash[id] += 1}
+      return counts.select {|_, val| val == assigned_sections.length}.keys
+    end
+  end
```
Comment (aoby) [inline dashboard/app/models/user.rb]: I think this might be even more readable if it's a single call to one or another private method based on user type, rather than one helper and the other inline:

Something like
```ruby
teacher? ? 
  get_teacher_hidden_ids(hidden_stages) : 
  get_student_hidden_ids(assign_id, hidden_stages)
```

File: dashboard/app/models/user.rb [hunk 16]
```diff
@@ -1563,4 +1577,67 @@ def associate_with_potential_pd_enrollments
...
+  # on the input. If hidden_stages is true, id is expected to be a script id and
+  # we look for stages that are hidden. If hidden_stages is false, id is expected
+  # to be a course_id, and we look for hidden scripts.
+  # @param {string} assign_id - Course/script id we're asking for hidden elements of
+  # @param {boolean} [hidden_stages] - Whether we're looking for hidden stages or scripts
+  # @return {Hash<string,number[]>|number[]}
+  #   For teachers, this will be a hash mapping from section id to a list of hidden
+  #   ids for that section.
+  #   For students this will just be a list of ids that are hidden for them.
+  def get_hidden_ids(assign_id, hidden_stages = true)
```
Comment (aoby) [inline dashboard/app/models/user.rb]: I didn't notice until this was moved to private that this method isn't called directly from outside.

It looks like the entry points are always explicitly through either `get_hidden_stage_ids` or `get_hidden_script_ids` so does it make sense to combine into one method just to split again based on the `hidden_stages` parameter?

Might it make more sense to keep the stage vs. script logic separate and converge on filtering the sections by user type and constructing the final array vs. hash?

Comment (Bjvanminnen) [reply dashboard/app/models/user.rb]: The main source of complexity here is in figuring out what set of sections to look at for students when determining which stages/scripts are hidden. As such, that's the bit I'd like to make sure I'm not repeating/introducing subtle differences to.

It might make sense however to get rid of `get_hidden_ids` and to have each of the callers instead just own that check, i.e.

```ruby
def get_hidden_stage_ids(script_name)
    script = Script.get_from_cache(script_name)
    return [] if script.nil?

    teacher? ? get_teacher_hidden_ids(true) : get_student_hidden_ids(script.id, true)
  end

  def get_hidden_script_ids(course)
    return [] if course.nil?

    teacher? ? get_teacher_hidden_ids(false) : get_student_hidden_ids(course.id, false)
  end
```

Thoughts?

Comment (aoby) [reply dashboard/app/models/user.rb]: I presume you meant to pass `false` in the 2nd case?

I like it. That seems cleaner. 
Since this will still return either a hash or array, I'd like to see that `@returns` comment on each of these (public) methods.

Is there a chance we'll want types other than stage or script in the future? Should this be an enum (e.g. hidden_id_type) instead of a bool?

Comment (Bjvanminnen) [reply dashboard/app/models/user.rb]: Yep, meant to have false in second case.

I don't think we'll be wanting other types in the future.
